subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ConnectionFailureRepositoryBlacklister.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.ivyservice.ivyresolve;
+
+import com.google.common.collect.Sets;
+import org.apache.commons.lang.exception.ExceptionUtils;
+import org.gradle.api.UncheckedIOException;
+
+import java.io.IOException;
+import java.util.Set;
+
+public class ConnectionFailureRepositoryBlacklister implements RepositoryBlacklister {
+
+    private final Set<String> blacklistedRepositories = Sets.newConcurrentHashSet();
```

Comment:
:o: I am not sure if I interpret the Javadocs of [ConcurrentHashSet](http://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html?is-external=true) properly here but it sounds like you still need to synchronize.

> Retrieval operations (including get) generally do not block, so may overlap with update operations (including put and remove). Retrievals reflect the results of the most recently completed update operations holding upon their onset.

We should probably add some concurrency test for this class.

Reply:
What the JavaDoc says is that a `get` may not see a very recent `add`. Since we're only talking about a few CPU cycles here and the method is called from long-running operations (repository access), it is highly unlikely that this will ever happen. If it does happen, there is no loss of correctness, we just try to hit the broken repo one more time.

By synchronizing we would optimize for this rare corner case, but make the usual case slower.   I think the concurrent set is the right tool for the job here.

Reply:
Things are much worse than the incorrectness. `HashSet` and `HashMap` are not thread-safe and concurrent write operation might cause infinite loops in very rare cases. Actually I had met 100% CPU usage several times previously.  https://bugs.openjdk.java.net/browse/JDK-6423457